Model/PaypalExpress/QuoteCreate.php

```diff
@@ -0,0 +1,292 @@
+        CustomerRepositoryInterface $customerRepository,
+        CheckoutSession $checkoutSession,
+        QuoteRepository $quoteRepository,
+        ShipmentEstimationInterface $shipmentEstimation,
+        Log $logger
+    ) {
+        $this->responseFactory = $responseFactory;
+        $this->quoteBuilderInterfaceFactory = $quoteBuilderInterfaceFactory;
+        $this->customerSession = $customerSession;
+        $this->customerRepository = $customerRepository;
+        $this->checkoutSession = $checkoutSession;
+        $this->quoteRepository = $quoteRepository;
+        $this->shipmentEstimation = $shipmentEstimation;
+        $this->logger = $logger;
+    }
+    /** @inheritDoc */
+    public function execute(
+        ShippingAddressRequestInterface $shipping_address,
+        string $page,
+        $form_data = null
+    ) {
+
+        if ($page === 'product') {
+            $this->validateData($form_data);
+            $this->quote = $this->createQuote($form_data);
+        } else {
+            $this->quote = $this->checkoutSession->getQuote();
+        }
+
+        try {
+            $this->addAddressToQuote($shipping_address);
+            $this->setPaymentMethod();
+        } catch (\Throwable $th) {
+            $this->logger->addDebug(__METHOD__.$th->getMessage());
+            throw new PaypalExpressException(__("Failed to add address quote"), 1, $th);
+        }
+
+
+        $this->calculateQuoteTotals();
+        return $this->responseFactory->create(["quote" => $this->quote]);
+    }
+
+    /**
+     * Calculate quote totals, set store id required for quote masking,
+     * set customer email required for order validation
+     * @return void
+     */
+    protected function calculateQuoteTotals()
+    {
+        $this->quote->setStoreId($this->quote->getStore()->getId());   
+        
+        if ($this->quote->getCustomerEmail() === null) {
+            $this->quote->setCustomerEmail('no-reply@example.com');
+        }
+        $this->quote
+            ->setTotalsCollectedFlag(false)
+            ->collectTotals();
+
+        $this->quoteRepository->save($this->quote);
+    }
+    /**
+     * Add address from paypal express to quote
+     *
+     * @param ShippingAddressRequestInterface $shipping_address
+     */
+    protected function addAddressToQuote(ShippingAddressRequestInterface $shipping_address)
+    {
+        if ($this->customerSession->isLoggedIn()) {
+            $this->quote->assignCustomerWithAddressChange(
+                $this->customerRepository->getById($this->customerSession->getCustomerId())
+            );
+        }
+
+        $address = $this->quote->getShippingAddress();
+
+        $address->setCountryId($shipping_address->getCountryCode());
+        $address->setPostcode($shipping_address->getPostalCode());
+        $address->setCity($shipping_address->getCity());
+        $address->setRegion($shipping_address->getState());
+        $this->maybeFillAnyMissingAddressFields($shipping_address);
```

Comment:
This should be discussed - as residual values would stay and that can impact delivery of the address...

Reply:
from what I see in m2 the calculation is made based on 4 address fields: postal code, city, region and street, all 4 are updated by this code